src/core/ComposableElement.php

```diff
@@ -466,31 +466,8 @@ protected function init(): void {}
       for ($i = 0; $i < $ln; ++$i) {
         $child = $this->children->get($i);
         if ($child instanceof :x:element) {
-          do {
-            assert($child instanceof :x:element);
-            if ($child instanceof XHPAwaitable) {
-              $child = static::__xhpAsyncRender($child)->getWaitHandle();
-            } else {
-              $child = $child->render();
-            }
-            if ($child instanceof WaitHandle) {
-              $childWaitHandles[$i] = $child;
-            } else if ($child instanceof :x:element) {
-              continue;
-            } else if ($child instanceof :x:frag) {
-              $children = $this->children->toValuesArray();
-              array_splice($children, $i, 1, $child->getChildren());
-              $this->children = new Vector($children);
-              $ln = count($this->children);
-              --$i;
-            } else if ($child === null) {
-              $this->children->removeKey($i);
-              $i--;
-            } else {
-              assert($child instanceof XHPChild);
-              $this->children[$i] = $child;
-            }
-          } while ($child instanceof :x:element);
+          $child = $child->__renderAndProcess()->getWaitHandle();
+          $childWaitHandles[$i] = $child;
```

Comment:
Deleted code:

- null is not a valid result from :x:element since 2.0
- the other cases, including :x:frag were already handled in almost-duplicate code for XHPAsync from ~ L454
- with this new approach, we always have a waithandle, so the code for handling XHPAsync is always used

I think the block ~ L454 should still handle parallelization - will add some tests for this.